add post previewing postprocessing (#708)
tidyverse/tidyverse.org: content/blog/postprocessing-preview/index.Rmd

---
output: hugodown::hugo_document

slug: postprocessing-preview
title: Postprocessing is coming to tidymodels

> date: 2024-10-01

Author: TODO: bump this!

Postprocessing support with tidymodels hasn't yet made it to CRAN, but you can install the needed versions of tidymodels packages with the following code.

```{r}
#| label: install-packages
#| eval: false
pak::pak(
  paste0(
    "tidymodels/",
    c("tune", "workflows", "rsample", "tailor")
  )
)
```

Now, we load packages with those developmental versions installed.

```{r}
#| label: load-packages
#| message: false
library(tidymodels)
library(probably)
library(tailor)
```

Existing tidymodels users might have spotted something funky already; who is this tailor character?

## Meet tailor👋

The tailor package introduces tailor objects, which compose iterative adjustments to model predictions. tailor is to postprocessing as recipes is to preprocessing; applying your mental model of recipes to tailor should get you a good bit of the way there.

> | Tool | Applied to\... | Initialize with\... | Composes\... | Train with\... | Predict with\... |

Author: Good thought--b724494!

tailor()
```

Tailors compose "adjustments," analogous to steps from the recipes package.

```{r}
#| label: tailor-adjust-1
tailor() %>%
  adjust_probability_threshold(threshold = .7)
```

As an example, we'll apply this tailor to the `two_class_example` data made available after loading tidymodels.

```{r}
#| label: two-class-example
head(two_class_example)
```

This data gives the true value of an outcome variable `truth` as well as predicted probabilities (`Class1` and `Class2`). The hard class predictions, in predicted, are `"Class1"` if the probability assigned to `"Class1"` is above .5, and `"Class2"` otherwise.

The model predicts `"Class1"` more often than it does `"Class2"`.

```{r}
#| label: count-two-class-example
two_class_example %>% count(predicted)
```

If we wanted the model to predict `"Class2"` more often, we could increase the probability threshold assigned to `"Class1"` above which the hard class prediction will be `"Class1"`. In the tailor package, this adjustment is implemented in `adjust_probability_threshold()`, which can be situated in a tailor object.

```{r}
#| label: tlr
tlr <-
  tailor() %>%
  adjust_probability_threshold(threshold = .7)

tlr
```

> tailors must be fitted before they can predict on new data. For adjustments like `adjust_probability_threshold()`, there's no training that actually happens at the `fit()` step besides recording the name and type of relevant variables. For other adjustments, like numeric calibration with `adjust_numeric_calibration()`, parameters are actually estimated at the `fit()` step and separate data should be used to train the postprocessor and evaluate its performance. More on this in [Tailors in context](#tailors-in-context).

Reviewer: just to avoid confusion since we mentioned recipes and steps

```suggestion
tailors must be fitted before they can predict on new data. For adjustments like `adjust_probability_threshold()`, there's no training that actually happens at the `fit()` step besides recording the name and type of relevant variables. For other adjustments, like numeric calibration with `adjust_numeric_calibration()`, parameters are actually estimated at the `fit()` stage and separate data should be used to train the postprocessor and evaluate its performance. More on this in [Tailors in context](#tailors-in-context).
```

Now, when passed new data, the trained tailor will determine the outputted class based on whether the probability assigned to the level `"Class1"` is above `.7`, resulting in more predictions of `"Class2"` than before.

```{r}
#| label: count-tlr-trained
predict(tlr_trained, two_class_example) %>% count(predicted)
```

Changing the probability threshold is one of many possible adjustments available in tailor.

-   For probabilities: [calibration](https://tailor.tidymodels.org/reference/adjust_probability_calibration.html)
-   For transformation of probabilities to hard class predictions: [thresholds](https://tailor.tidymodels.org/reference/adjust_probability_threshold.html), [equivocal zones](https://tailor.tidymodels.org/reference/adjust_equivocal_zone.html)
-   For numeric outcomes: [calibration](https://tailor.tidymodels.org/reference/adjust_numeric_calibration.html), [range](https://tailor.tidymodels.org/reference/adjust_numeric_range.html)

> Support for tailors in now plumbed through workflows (via [`add_tailor()`](https://workflows.tidymodels.org/dev/reference/add_tailor.html)) and tune, and rsample includes a set of infastructural changes to prevent data leakage behind the scenes. That said, we haven't yet implemented support for tuning parameters in tailors, but we plan to implement that before this functionality heads to CRAN.

Reviewer: ```suggestion
Support for tailors is now plumbed through workflows (via [`add_tailor()`](https://workflows.tidymodels.org/dev/reference/add_tailor.html)) and tune, and rsample includes a set of infrastructural changes to prevent data leakage behind the scenes. That said, we haven't yet implemented support for tuning parameters in tailors, but we plan to implement that before this functionality heads to CRAN.
```

```{r}
#| label: deliveries
data(deliveries)

# split into training and testing sets
set.seed(1)
delivery_split <- initial_split(deliveries)
delivery_train <- training(delivery_split)
delivery_test  <- testing(delivery_split)

# resample the training set using 10-fold cross-validation
set.seed(1)
delivery_folds <- vfold_cv(delivery_train)

# print out the training set
delivery_train
```

Let's deliberately define a regression model that has poor predicted values: a boosted tree with only three ensemble members.

```{r}
#| label: bad-boost
delivery_wflow <-
  workflow() %>%
  add_formula(time_to_delivery ~ .) %>%
  add_model(boost_tree(mode = "regression", trees = 3))
```

Evaluating against resamples:

```{r}
#| label: resample-bad-boost
set.seed(1)
delivery_res <- 
  fit_resamples(
    delivery_wflow, 
    delivery_folds, 
    control = control_resamples(save_pred = TRUE)
  )
```

The $R^2$ looks quite strong!

```{r}
#| label: metrics-bad-boost
collect_metrics(delivery_res)
```

> Let's take a closer look at the predictions, though. How well is it calibrated? We can use the `cal_plot_regression()` helper from the probably package to put together a quick diagnostic plot.

Author: ```suggestion
Let's take a closer look at the predictions, though. How well are they calibrated? We can use the `cal_plot_regression()` helper from the probably package to put together a quick diagnostic plot.
```

```

There's actually some tricky data leakage prevention happening under the hood here. When you add tailors to workflow and fit them with tune, this is all taken care of for you. If you're interested in using tailors outside of that context, check out [this documentation section](https://workflows.tidymodels.org/dev/reference/add_tailor.html#data-usage) in `add_tailor()`.

## What's to come

We're excited about how this work is shaping up and would love to hear yall's thoughts on what we've brought together so far. Please do comment on our social media posts about this blog entry or leave issues on the [tailor GitHub repository](https://github.com/tidymodels/tailor) and let us know what you think!

Before these changes head out to CRAN, we'll also be implementing tuning functionality for postprocessors. You'll be able to tag arguments like `adjust_probability_threshold(threshold)` or `adjust_probability_calibration(method)` with `tune()` to optimize across several values. Besides that, post-processing with tidymodels should "just work" on the developmental versions of our packages---let us know if you come across anything wonky.

> 

Author: I think introducing support for conformal prediction intervals or applicability domains into the same interfaces as post-processors leads to a few headaches for us and for users.

* If in tailor: The story that post-processors update predictions no longer holds up.
* If in workflows directly: Workflows, to this point, have one object per stage. (e.g. can have a recipe or a formula preprocessor but not both.) It's of course possible to undo this engineering-wise, but the story is again less clear.
* I don't imagine that users would want to calculate conformal prediction intervals or applicability domains at any point other than the final model fit? e.g. when resampling/tuning, this seems like a lot of computation for an intermediate fit that wouldn't inform model selection, from my understanding. This feels like to me like an argument to keep those operations living outside of the workflow.
* Isn't the result of an applicability domain independent of the model? i.e. a function of the distance between the training data and the data supplied for prediction and will be the same whether the model is a boosted tree or a linear model?

I'm game to discuss these features, but don't imagine a "yes" or "no" from users would greatly affect that discussion.
